trl/trainer/dpo_trainer.py

```diff
@@ -189,24 +202,46 @@ def concatenated_inputs(self, batch: Dict[str, Union[List, torch.LongTensor]]) -
-                concatenated_key = k.replace("rejected", "concatenated")
-                concatenated_batch[concatenated_key] = torch.cat(
-                    (
-                        concatenated_batch[concatenated_key],
-                        pad_to_length(batch[k], max_length, pad_value=pad_value),
-                    ),
-                    dim=0,
-                ).to(self.accelerator.device)
+        
+        if self.is_encoder_decoder:
+            max_length = max(batch["chosen_labels"].shape[1], batch["rejected_labels"].shape[1])
+            for k in batch:
+                if k.startswith("chosen") and isinstance(batch[k], torch.Tensor):
+                    pad_value = self.label_pad_token_id 
+                    concatenated_key = k.replace("chosen", "concatenated")
+                    concatenated_batch[concatenated_key] = pad_to_length(batch[k], max_length, pad_value=pad_value)
+            for k in batch:
+                if k.startswith("rejected") and isinstance(batch[k], torch.Tensor):
+                    pad_value = self.label_pad_token_id
+                    concatenated_key = k.replace("rejected", "concatenated")
+                    concatenated_batch[concatenated_key] = torch.cat(
+                        (
+                            concatenated_batch[concatenated_key],
+                            pad_to_length(batch[k], max_length, pad_value=pad_value),
+                        ),
+                        dim=0,
+                    ).to(self.accelerator.device)
+            concatenated_batch["concatenated_input_ids"] = batch["prompt_input_ids"].repeat(2,1)
+            concatenated_batch["concatenated_attention_mask"] = batch["prompt_attention_mask"].repeat(2,1)
+        else:
+            max_length = max(batch["chosen_input_ids"].shape[1], batch["rejected_input_ids"].shape[1])
+            for k in batch:
+                if k.startswith("chosen") and isinstance(batch[k], torch.Tensor):
+                    pad_value = self.label_pad_token_id if "labels" in k else self.padding_value
+                    concatenated_key = k.replace("chosen", "concatenated")
+                    concatenated_batch[concatenated_key] = pad_to_length(batch[k], max_length, pad_value=pad_value)
+            for k in batch:
+                if k.startswith("rejected") and isinstance(batch[k], torch.Tensor):
+                    pad_value = self.label_pad_token_id if "labels" in k else self.padding_value
+                    concatenated_key = k.replace("rejected", "concatenated")
+                    concatenated_batch[concatenated_key] = torch.cat(
+                        (
+                            concatenated_batch[concatenated_key],
+                            pad_to_length(batch[k], max_length, pad_value=pad_value),
+                        ),
+                        dim=0,
+                    ).to(self.accelerator.device)
```

Comment:
```suggestion
        if self.is_encoder_decoder:
            max_length = max(batch["chosen_labels"].shape[1], batch["rejected_labels"].shape[1])
            pad_value = self.label_pad_token_id 
        else:
            max_length = max(batch["chosen_input_ids"].shape[1], batch["rejected_input_ids"].shape[1])
            pad_value = self.label_pad_token_id if "labels" in k else self.padding_value

        for k in batch:
            if k.startswith("chosen") and isinstance(batch[k], torch.Tensor):
                concatenated_key = k.replace("chosen", "concatenated")
                concatenated_batch[concatenated_key] = pad_to_length(batch[k], max_length, pad_value=pad_value)
        for k in batch:
            if k.startswith("rejected") and isinstance(batch[k], torch.Tensor):
                concatenated_key = k.replace("rejected", "concatenated")
                concatenated_batch[concatenated_key] = torch.cat(
                    (
                        concatenated_batch[concatenated_key],
                        pad_to_length(batch[k], max_length, pad_value=pad_value),
                    ),
                    dim=0,
                ).to(self.accelerator.device)
            if self.is_encoder_decoder:
                concatenated_batch["concatenated_input_ids"] = batch["prompt_input_ids"].repeat(2,1)
                concatenated_batch["concatenated_attention_mask"] = batch["prompt_attention_mask"].repeat(2,1)
```
I believe this is equivalent.

Reply:
I don't think this will work as k is not defined at this point for the pad_value

Reply:
Oh, whoops. I missed that variable.
That said, the pad_value can be moved down somewhat in my solution, and it'll still remove some duplicated code.

Reply:
Yes I agree